pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerInterceptors.java

```diff
@@ -109,9 +110,29 @@ public void onAcknowledge(Consumer<T> consumer, MessageId messageId, Throwable e
      * @param exception exception returned by broker.
      */
     public void onAcknowledgeCumulative(Consumer<T> consumer, MessageId messageId, Throwable exception) {
-        for (int i = 0; i < interceptors.size(); i++) {
+        for (ConsumerInterceptor<T> interceptor : interceptors) {
             try {
-                interceptors.get(i).onAcknowledgeCumulative(consumer, messageId, exception);
+                interceptor.onAcknowledgeCumulative(consumer, messageId, exception);
+            } catch (Exception e) {
+                log.warn("Error executing interceptor onAcknowledgeCumulative callback ", e);
+            }
+        }
+    }
+
+    /**
+     * This method will be called when a redelivery from a negative acknowledge occurs.
+     * <p>
+     * This method calls {@link ConsumerInterceptor#onNegativeAckRedelivery(Consumer, Set<MessageId>)} method for each interceptor.
+     * <p>
+     * This method does not throw exceptions. Exceptions thrown by any of interceptors in the chain are logged, but not propagated.
+     *
+     * @param consumer the consumer which contains the interceptors
+     * @param messageIds set of messages ids being redelivery.
+     */
+    public void onNegativeAckRedelivery(Consumer<T> consumer, Set<MessageId> messageIds) {
```

Comment:
Regarding the naming of this method, I think it might get a bit misleading because the redelivery might actually get to a different consumer. Eg. ConsumerA calls Nack() and after 1 min ConsumerB will get the message again.

Would it be more intuitive to have the method called `onNegativeAck()` and call it as soon as the app calls `consumer.negativeAcknowledgement()`?

Reply:
Yep, at a first glance I thought exactly that, but later I realise that calling `onNegativeAck()` when `consumer.negativeAcknowledgement()` will occur almost at that very moment (`consumer.negativeAcknowledgement()` just adds the msg to `negativeAckTracker`) and the user could solve this by himself, instead, with adding `onNegativeAckRedelivery()` offers a feature which the user cannot solve right now.

The good thing about `onNegativeAck()` is that is consistent with current `onAcknowledge()` existing method.

Maybe we want to add both features? and in some future also adding the proper `on*()` method for `UnackMessageTracker`?

Reply:
Or maybe we can rename this pr in favor of `onNegativeAck()` and add in a different one the `onNegativeAckRedelivery()` feature if you want.

Reply:
> when consumer.negativeAcknowledgement() will occur almost at that very moment (consumer.negativeAcknowledgement() just adds the msg to negativeAckTracker) and the user could solve this by himself, instead, with adding onNegativeAckRedelivery() offers a feature which the user cannot solve right now.

Reply:
`onNegativeAcksSend` sounds good to me :+1: , what I like from *redelivery* is that express its implementation behaviour, but this is probably biased by my perception of knowing how is implemented.

Do I proceed with the mentioned rename? @merlimat @codelipenghui

Reply:
👍

Reply:
Done! Thanks for the feedback.